File: 0076-minimum-window-substring/0076-minimum-window-substring.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> targetChars;
    for (char c : t) {
        targetChars[c]++;
    }

    int left = 0;
    int minLen = INT_MAX;
    int minStart = 0;
    int requiredChars = targetChars.size();

    for (int right = 0; right < s.length(); right++) {
        char rightChar = s[right];
        if (targetChars.count(rightChar)) {
            targetChars[rightChar]--;
            if (targetChars[rightChar] == 0) {
                requiredChars--;
            }
        }

        while (requiredChars == 0) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                minStart = left;
            }

            char leftChar = s[left];
            if (targetChars.count(leftChar)) {
                targetChars[leftChar]++;
                if (targetChars[leftChar] > 0) {
                    requiredChars++;
                }
            }

            left++;
        }
    }

    return minLen == INT_MAX ? "" : s.substr(minStart, minLen);
    }
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        Map<Character, Integer> tFreq = new HashMap<>();
        for(char c: t.toCharArray()){
            tFreq.put(c, tFreq.getOrDefault(c, 0)+1);
        }
        
        int left = 0;
        int required = tFreq.size();
        int minLen = Integer.MAX_VALUE;
        String result = "";
        
        for(int right=0; right<s.length(); right++){
            if(tFreq.containsKey(s.charAt(right))){
                tFreq.put(s.charAt(right), tFreq.get(s.charAt(right))-1);
                if(tFreq.get(s.charAt(right)) == 0){
                    required--;
                }
            }
            
            while(required == 0){
                if(right-left+1 < minLen){
                    minLen = right-left+1;
                    result = s.substring(left, right+1);
                    //minStart = left;
                }
                
                if(tFreq.containsKey(s.charAt(left))){
                    tFreq.put(s.charAt(left), tFreq.get(s.charAt(left))+1);
                    if(tFreq.get(s.charAt(left)) > 0){
                        required++;
                    }
                }
                left++;
            }
        }
        
        return minLen == Integer.MAX_VALUE? "" :  result;
    }
}